### DISCUSSION TITLE
How to add support for a new protocol?

### DISCUSSION THREAD

**Discussion by jonaslagoni** [microcks/microcks#702] [Q&A] [answered]
Hey!

I would love to try out Microcks on my [NATS](https://nats.io/) setup with AsyncAPI, but as far as I can tell, it's not possible to mock with the NATS protocol (or have I simply missed it?).

So my question is, where can I find a guide on how to add support for it?

**Comment by lbroudoux** [comment, marked as answer]
Hi Jonas!

You're right: as of today, we do not support NATS mocking - but agreed that it would be a nice addition.
We do not have a guide on how to add a new broker/connector type at the moment.

Basically, adding a new async protocol can be split in 3 different tasks:
* Adding detection/parsing of protocol binding in AsyncAPIImporter (see https://github.com/microcks/microcks/blob/master/webapp/src/main/java/io/github/microcks/util/asyncapi/AsyncAPIImporter.java#L378 as an example for AMQP/RabbitMQ)
* Adding a new ProducerManager implementation for the mocking part in async-minion (see https://github.com/microcks/microcks/blob/master/minions/async/src/main/java/io/github/microcks/minion/async/producer/AMQPProducerManager.java as an example)
* Adding a new MessageConsumptionTask implementation for the testing part in async-minion (see https://github.com/microcks/microcks/blob/master/minions/async/src/main/java/io/github/microcks/minion/async/consumer/AMQPMessageConsumptionTask.java as an example)

Let me know what you think 😉

**Reply by jonaslagoni** [reply]
Thanks for the quick reply! Gonna try it and probably write a contribution guide alongside the feature then 😄

**Reply by lbroudoux** [reply]
Awesome! Let me know if you need any help or more details, we could set up a quick call.
